internal/exec/util/selabeling.c

```diff
@@ -0,0 +1,160 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#define _GNU_SOURCE
+#include <assert.h>
+#include <errno.h>
+#include <limits.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdarg.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/xattr.h>
+#include <unistd.h>
+
+#include <selinux/selinux.h>
+#include <selinux/label.h>
+
+#include "selabeling.h"
+
+/* The main reason we do this from C here instead of through cgo (apart from the
+ * latter being painful to work with) is for error-handling: we're using the
+ * APIs which set errno, which we simply cannot access from Go:
+ * - https://github.com/golang/go/issues/1360
+ * - https://codereview.appspot.com/3880041/
+ *
+ * Relatedly, we construct a good error string here instead of returning a
+ * result_t style enum like in `blkid.c`. We could combine the two in the future
+ * if we need the Go side to special-case an error, though today these errors
+ * are not really actionable.
+ */
+
+struct selabeling_context {
+    char *root;
+    struct selabel_handle *handle;
+};
+
+static char*
+malloc_vprintf(const char *fmt, va_list args)
+{
+  /* see vsnprintf(3) EXAMPLE */
+  va_list argsc;
+  va_copy(argsc, args);
+  int n = vsnprintf(NULL, 0, fmt, argsc);
+  va_end(argsc);
+  assert(n > 0);
+  n++; /* for final NUL */
+  char *s = malloc(n);
+  assert(vsnprintf(s, n, fmt, args) == n-1);
+  return s;
+}
+
+static char*
+malloc_sprintf(const char *fmt, ...)
+{
+  va_list args;
+  va_start (args, fmt);
+  char *s = malloc_vprintf(fmt, args);
+  va_end (args);
+  return s;
+}
+
+/* see similar UX function glnx_throw_errno_prefix() in libglnx */
+static int
+throw_errno_prefix(char **error, const char *fmt, ...)
+{
+  if (!error)
+    return -1;
+
+  int errsv = errno;
+  va_list args;
+  va_start (args, fmt);
+
+  char *prefix = malloc_vprintf(fmt, args);
+  char buf[1024]; /* see NOTES in strerror(3) */
+  *error = malloc_sprintf("%s: %s", prefix, strerror_r(errsv, buf, sizeof(buf)));
+
+  va_end (args);
+  errno = errsv;
+  return -1;
+}
+
+int
+selabeling_new(const char *root, struct selabeling_context **out_ctx, char **error)
+{
+  assert(out_ctx);
+
+  /* XXX: actually read from /etc/selinux/config the policy name; probably cleaner to do
+   * this from the Go side and pass in the policy name. */
+  char policy_root[PATH_MAX];
+  int n = snprintf(policy_root, sizeof(policy_root), "%s/etc/selinux/targeted", root);
+  assert(n > 0 && n < PATH_MAX);
+
+  if (selinux_set_policy_root(policy_root) != 0)
+    return throw_errno_prefix(error, "setting policy root to %s", policy_root);
+
+  struct selabel_handle *handle = selabel_open(SELABEL_CTX_FILE, NULL, 0);
+  if (!handle)
+    return throw_errno_prefix(error, "initializing file context backend handle");
+
+  *out_ctx = malloc(sizeof(struct selabeling_context));
```

Comment:
Let's always use `xmalloc` or so so we cleanly abort on OOM.

Reply:
Hmm, doesn't seem like `xmalloc` is in the default glibc headers? Anyway, since it's the only spot we `malloc`, I added an `assert` there.

Reply:
Right, `xmalloc` is a gnulib invention, GLib's `g_malloc()` also similarly aborts on OOM, so does most of the Rust stdlib, etc.  We could cargo-cult in the gnulib version, but eh, an assert is OK for now too.